ops/pebble.py

```diff
@@ -1926,3 +1910,188 @@ def _websocket_url(self, task_id: str, websocket_id: str) -> str:
+        else:
+            # Consume the boundary and get ready for parsing the headers.
+            self._buffer = self._buffer[span[1]:]
+            next_state = self._state_parse_header
+        return next_state
+
+    def _state_parse_header(self):
+        try:
+            headers, content = self._buffer.split(b'\r\n\r\n', 1)
+        except ValueError:
+            # Not enough data; stop for now
+            raise StopParsing()
+
+        # Rely on the stdlib for parsing the headers, since this can be done purely
+        # in memory without any realistic worry of memory concerns.
+        parser = email.parser.BytesFeedParser()
+        parser.feed(headers)
+        message = parser.close()
+
+        content_disposition = message.get_content_disposition()
+        if content_disposition != 'form-data':
+            raise RuntimeError(
+                'unexpected content disposition: {}'.format(content_disposition))
+
+        name = message.get_param('name', header='content-disposition')
+        if name == 'response':
+            next_state = self._state_parse_response_part
+        elif name == 'files':
+            filename = message.get_filename()
+            if filename is None:
+                raise NotImplementedError('multipart "files" part missing filename')
+            self._prepare_tempfile(filename)
+            next_state = self._state_parse_files_part
+        else:
+            raise RuntimeError(
+                'unexpected name in content-disposition header: {}'.format(name))
+
+        # Consume the header; now we're ready to parse the part body
+        # (either "request" or "files")
+        self._buffer = content
+        return next_state
+
+    def _state_parse_response_part(self):
+        # Response is handled purely in memory; wait on parsing until we've received
+        # the full part.
+        match = self._boundary_rgx.search(self._buffer)
+        if match is None:
+            raise StopParsing()
+
+        span = match.span()
+        # Content is terminated by \r\n<boundary>; adjust for the \r\n
+        end_of_content = span[0] - 2
+        content = self._buffer[:end_of_content]
+
+        self._response = json.loads(content.decode())
+
+        self._buffer = self._buffer[span[0]:]
+        return self._state_detect_boundary
+
+    def _state_parse_files_part(self):
+        # Files can be huge; we shouldn't keep them in memory; rather we should
+        # stream them to files.  For now, using tempfiles.
+
+        match = self._boundary_rgx.search(self._buffer)
+        if match is None:
+            # We may have data, but we haven't reached the end of the file.
+            # Store as much data as we safely can without accidentally storing
+            # bytes of the boundary line.
+            content = self._buffer[:-self._safe_buffer_len]
+            self._buffer = self._buffer[-self._safe_buffer_len:]
+            with self._open_tempfile() as outfile:
```

Comment:
It seems inefficient to close and reopen the temp file every time we need to write to it (when we're only processing one file at a time). Can't we keep it open and just write to the open file?